# 01. Importing Libraries

In [2]:
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [4]:

# Create path
path = r'/Users/muhammaddildar/Desktop/03-2025 Instacart Basket Analysis/Data'

# Orders data
df_ords = pd.read_csv(os.path.join(path, 'Original Data', 'orders.csv'), index_col=False)

# Products data
df_prods = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'), index_col=False)

# Departments data
df_dep = pd.read_csv(os.path.join(path, 'Original Data', 'departments.csv'), index_col=False)


# 03. Wrangling Data

## Departments data frame

In [7]:
#transposing df_dep
df_dep_t=df_dep.T

## Changing header of departments df

In [9]:
#take 1st row of df_dep_t for the header
new_header = df_dep_t.iloc[0]

In [10]:
#copying df_dep_t from row 1 onward
df_dep_t_new = df_dep_t[1:]

In [11]:
#set the header row as the df header
df_dep_t_new.columns = new_header


## Transformation result

In [13]:
df_dep_t_new

department_id,department
1,frozen
2,other
3,bakery
4,produce
5,alcohol
6,international
7,beverages
8,pets
9,dry goods pasta
10,bulk


# 04. Creating Data Dictionary

In [15]:
data_dict = df_dep_t_new.to_dict('index')

# 05. TASK

In [17]:
# Step 2: Since the order number serves as a unique identifier for each user's order, it's more appropriate to treat it as a string rather than a numeric value.
df_ords['order_number'] = df_ords['order_number'].astype('str')

In [18]:
df_ords.describe()

,order_id,user_id,order_dow,order_hour_of_day,days_since_prior_order
count,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.214874e+06
mean,1.710542e+06,1.029782e+05,2.776219e+00,1.345202e+01,1.111484e+01
std,9.875817e+05,5.953372e+04,2.046829e+00,4.226088e+00,9.206737e+00
min,1.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,8.552715e+05,5.139400e+04,1.000000e+00,1.000000e+01,4.000000e+00
50%,1.710542e+06,1.026890e+05,3.000000e+00,1.300000e+01,7.000000e+00
75%,2.565812e+06,1.543850e+05,5.000000e+00,1.600000e+01,1.500000e+01
max,3.421083e+06,2.062090e+05,6.000000e+00,2.300000e+01,3.000000e+01


In [19]:
# Renaming the column without overwriting the original DataFrame
df_ords_renamed = df_ords.rename(columns={'order_dow': 'orders_day_of_the_week'}, inplace=False)


In [20]:
df_ords_renamed

,order_id,user_id,eval_set,order_number,orders_day_of_the_week,order_hour_of_day,days_since_prior_order
0,2539329,1,prior,1,2,8,NaN
1,2398795,1,prior,2,3,7,15.0
2,473747,1,prior,3,3,12,21.0
3,2254736,1,prior,4,4,7,29.0
4,431534,1,prior,5,4,15,28.0
...,...,...,...,...,...,...,...
3421078,2266710,206209,prior,10,5,18,29.0
3421079,1854736,206209,prior,11,4,10,30.0
3421080,626363,206209,prior,12,1,12,18.0
3421081,2977660,206209,prior,13,1,12,7.0


In [21]:
#Step 4: Finding busiest hour
df_ords['order_hour_of_day'].value_counts(dropna = True)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

### Observation: 
The busiest hour of the day for orders is 10 AM, with a total of 288,418 orders. This indicates a peak in customer activity during this time.

In [23]:
#Step 5: What "4" means in the data dictionary
print(data_dict.get('4'))

{'department': 'produce'}


In [24]:
#Step 6: Creating subset for breakfast items. First, identify breakfast dept id:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [25]:
#creating the subset
df_prods_breakfast = df_prods.loc[df_prods['department_id'] == 14]

In [26]:
#checking result
df_prods_breakfast.head()

,product_id,product_name,aisle_id,department_id,prices
27,28,Wheat Chex Cereal,121,14,10.1
33,34,NaN,121,14,12.2
67,68,"Pancake Mix, Buttermilk",130,14,13.7
89,90,Smorz Cereal,121,14,3.9
210,211,Gluten Free Organic Cereal Coconut Maple Vanilla,130,14,3.6


In [27]:
#Step 7: creating "dinner party" subset.
df_prods_dinner_party = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

### Observation:
The sales team can use the data frame df_prods_dinner_party for their analysis.

In [29]:
#Step 8: find the number of rows
df_prods_dinner_party.shape

(7650, 5)

### Observation:
The new dinner party data frame has 7650 rows.

In [31]:
# Extract all the information for user_id == 1
df_user1 = df_ords[df_ords['user_id'] == 1]
df_user1


,order_id,user_id,eval_set,order_number,order_dow,order_hour_of_day,days_since_prior_order
0,2539329,1,prior,1,2,8,NaN
1,2398795,1,prior,2,3,7,15.0
2,473747,1,prior,3,3,12,21.0
3,2254736,1,prior,4,4,7,29.0
4,431534,1,prior,5,4,15,28.0
5,3367565,1,prior,6,2,7,19.0
6,550135,1,prior,7,1,9,20.0
7,3108588,1,prior,8,1,14,14.0
8,2295261,1,prior,9,1,16,0.0
9,2550362,1,prior,10,4,8,30.0


In [32]:
#Step 10: stats about the user's behaviour
df_user1.describe()

,order_id,user_id,order_dow,order_hour_of_day,days_since_prior_order
count,1.100000e+01,11.0,11.000000,11.000000,10.000000
mean,1.923450e+06,1.0,2.636364,10.090909,19.000000
std,1.071950e+06,0.0,1.286291,3.477198,9.030811
min,4.315340e+05,1.0,1.000000,7.000000,0.000000
25%,8.690170e+05,1.0,1.500000,7.500000,14.250000
50%,2.295261e+06,1.0,3.000000,8.000000,19.500000
75%,2.544846e+06,1.0,4.000000,13.000000,26.250000
max,3.367565e+06,1.0,4.000000,16.000000,30.000000


### Observation
This user exclusively places orders between Sunday and Wednesday, avoiding Thursday through Saturday. Over 50% of the time, their orders are placed by 8 AM or earlier. On average, they place orders approximately every 19 days. They never go more than 30 days without making an order and have placed two orders on the same day.

# 06. Export Data

In [35]:

### df_ords.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_wrangled.csv'))

# python
# Correct path for saving the CSV file
path = r'/Users/muhammaddildar/Desktop/03-2025 Instacart Basket Analysis/Data'

# Exporting the DataFrame to a CSV file (No index argument)
df_ords.to_csv(os.path.join(path,'Prepared Data', 'orders_wrangled.csv'))


In [38]:

### df_prods.to_csv(os.path.join(path, ' Data', 'Prepared Data', 'department_wrangled.csv'))

#python
# Correct path for saving the CSV file
path = r'/Users/muhammaddildar/Desktop/03-2025 Instacart Basket Analysis/Data'

# Exporting the DataFrame to a CSV file (No index argument)
df_dep.to_csv(os.path.join(path,'Prepared Data', 'department_wrangled.csv'))
